cognite/client/utils/_token_generator.py

```diff
@@ -29,15 +32,16 @@ def __init__(
             self._access_token_expires_at = None
 
     def return_access_token(self) -> str:
-        if not self.token_params_set():
-            raise CogniteAPIKeyError("Could not generate access token - missing token generation arguments")
-        elif self._access_token is None:
-            raise CogniteAPIKeyError("Could not generate access token from provided token generation arguments")
+        with LOCK:
+            if not self.token_params_set():
+                raise CogniteAPIKeyError("Could not generate access token - missing token generation arguments")
+            elif self._access_token is None:
+                raise CogniteAPIKeyError("Could not generate access token from provided token generation arguments")
 
-        if self._access_token_expires_at < datetime.datetime.now().timestamp():
-            self._generate_access_token()
+            if self._access_token_expires_at < datetime.datetime.now().timestamp() - 10:
```

Comment:
I'm probably missing something. Why is this buffer necessary?

Reply:
In my mind (at least 😅 ) there is a race condition here; since the thread can pass this check a few milliseconds before expiry, then not get de-scheduled by the os (is that a word?) and not run before after expiry. Maybe these tokens already have this kind of buffer "built in" (I am absolutely not familiar with "auth" in general)